### ISSUE TITLE
Tracking issue for RFC 2345, "Allow panicking in constants" (const_panic)

### ISSUE THREAD

**Issue by Centril** [rust-lang/rust#51999]
This is a tracking issue for the RFC *"Allow panicking in constants"* (rust-lang/rfcs#2345).

**Steps:**

- [x] Implement the RFC (cc @rust-lang/compiler -- can anyone write up mentoring instructions?)
- [ ] Adjust documentation ([see instructions on forge][doc-guide])
- [ ] Stabilization PR ([see instructions on forge][stabilization-guide])

[stabilization-guide]: https://forge.rust-lang.org/stabilization-guide.html
[doc-guide]: https://forge.rust-lang.org/stabilization-guide.html#updating-documentation

**Unresolved questions:**

- [ ] Should there be some additional message in the error about this being a panic turned error?
   Or do we just produce the exact message the panic would produce?

- [ ] This change becomes really useful if `Result::unwrap` and `Option::unwrap` become `const fn`, doing both in one go might be a good idea.

**Blockers:**

* ICE: https://github.com/rust-lang/rust/issues/66693

**Comment by clarcharr** [comment]
Don't forget `expect` and all the other unwrap-like methods. :P

**Comment by vi** [comment]
I suppose `unimplemented!()` and `unreachable!()` should also be included?

**Comment by SimonSapin** [comment]
I think they would be "automatically", since macros don’t have a separation similar to `fn` v.s. `const fn`.

**Comment by oli-obk** [comment]
I'll add some tests to #52011, but yes, it just works.

**Comment by oli-obk** [comment]
Yes

**Comment by IsaacWoods** [comment]
Now this is implemented, is there anything stopping `Option::unwrap` and friends from becoming `const fn`s (I'm assuming with a `rustc_const_unstable(feature = "??")`)?

**Comment by Centril** [comment]
@IsaacWoods to my knowledge there shouldn't be anything in the way of that.

**Comment by IsaacWoods** [comment]
@Centril great! I'll start work on a PR in that case

**Comment by clarcharr** [comment]
The thing in the way of that is running `fmt::Debug::fmt` in `const` context.

**Comment by oli-obk** [comment]
Yea, we need conditions in constants first

**Comment by Centril** [comment]
Right; bummer.

@oli-obk Any thoughts about the first unresolved question?

**Comment by oli-obk** [comment]
I'm for basically reporting it verbatim, but having the same prefix (`error[0080]:`) as all other errors.

**Comment by Centril** [comment]
@oli-obk I like that; (and it's a diagnostics issue anyways, not part of the spec, so it is up to T-compiler to decide this...).

**Comment by Centril** [comment]
@oli-obk I think sufficient time has passed for us to stabilize this... how do you feel about doing that? If you feel alright with that maybe you could write up a stabilization report?

**Comment by Centril** [comment]
@Nemo157 this basically permits `panic!("foo")` at the moment and not much more.

**Comment by oli-obk** [comment]
Yea, without control flow this feature is useless except for `unimplemented!()` working in const functions

**Comment by DoumanAsh** [comment]
Small question.
Would it allow to have functons like:

```rust
const fn static_assert(condition) {
    match condition {
        true => (),
        false => panic!("Assertion failed"),
     }
}
```

**Comment by oli-obk** [comment]
No, `match` is handled in https://github.com/rust-lang/rust/issues/49146

You can only do `const fn foo() { unimplemented!() }` as of right now.

**Comment by Swoorup** [comment]
Would this imply that const fn is not pure function?

**Comment by clarcharr** [comment]
Panicking doesn't cause impurity. The function will still always act the same for any given input.

**Comment by Swoorup** [comment]
I haven't seen any write-ups mentioning const fn would never evolve to be an impure function, so I am assuming you can't rely on this for purity?

**Comment by clarcharr** [comment]
I'm fairly certain that constants are by definition pure. What would it mean to have an impure constant?

**Comment by mark-i-m** [comment]
Imagine a `const fn rand_number_gen()` or something. All `const` means in my mind is that it runs at compile time. (That said, it's not clear to me how useful non-purity would be).

**Comment by clarcharr** [comment]
I feel like impure constants would be a bad idea, as it breaks determinism in builds. That said, I guess that this discussion isn't best had here.

**Comment by porky11** [comment]
@clarcharr 
It may be useful to run iterators at compile time. This would also require impure `const fn`.

**Comment by clarcharr** [comment]
@porky11 I think you're misunderstanding what impure means. Impure simply means that you get different values for the same input. Mutation isn't impure, as long as it's explicit. If you're passing a mutable reference as an argument (in this case, `&mut self`) you can think of it as treating the iterator state as both an input and an output. The same iterator state before will result in the same state after. Therefore, this is still pure.

Panicking is pure because it essentially ends the program when you reach a panic, and the circumstances leading up to it will always be the same. *Catching* panics could lead to impurity, but that's not mentioned at all here. I don't think that will ever be allowed.

**Comment by elichai** [comment]
Would be awesome to see compile time `assert(condition)` :)

**Comment by clarfon** [comment]
I believe that this will be possible with the effects traits RFC? On my phone right now but someone can probably share a link.

**Comment by elichai** [comment]
I actually found out(through people in the IRC) that it's already kind of possible if you cast the bool to usize and use it in indexing `&[()][1 - (condition as usize)]`

**Comment by mark-i-m** [comment]
Yes, but IIUC it produces inefficient code.

**Comment by clarfon** [comment]
Yeah, there are ways people do compile time assertions but the main issue is that there's no compile-time `Debug`.

**Comment by elichai** [comment]
No code to produce in compile time operations that result in ZST :)
https://godbolt.org/z/1sxuS-

**Comment by RalfJung** [comment]
Is there any forwards compatibility concern with eventually being able to *catch* panics in `const fn`?

**Comment by oli-obk** [comment]
I think only in the sense that users may expect panics in const eval to cause an error. So if you pass your function f to another function g to be called there, if f is called and panics, you may expect compilation to fail and it now isn't guaranteed, because g may catch the panic.

**Comment by oli-obk** [comment]
And we'd have this problem already for index out of bound panics, so there's nothing new

**Comment by RalfJung** [comment]
Okay. Our current implementation immediately aborts execution on a panic, we'd have to do something more clever if the panic could be caught... but I assume that should not be impossible.

**Comment by Ixrec** [comment]
Are there plans to allow catching panics in CTFE? I thought that feature was only useful/intended for keeping panics from unwinding into FFI, and the RFC doesn't even mention it.

**Comment by RalfJung** [comment]
Well I don't see why we'd not eventually permit catching panics in CTFE, given that we plan to allow pretty much everything that doesn't break determinism. ;)

**Comment by golddranks** [comment]
If everything goes well, if and match in constants are going to be released in 1.45 (https://github.com/rust-lang/rust/issues/49146). Is there any significant movement around this issue? I see the if/match and panics in constants as a valuable combination: together they bring in the ability for libraries to define const asserts and thus have better story around compile-time guarantees and library-level compiler errors.

**Comment by oli-obk** [comment]
Well, there are two unresolved questions and we need to document this feature. I don't see any actual problems with it though.

**Comment by golddranks** [comment]
My two cents about the unresolved questions. Hopefully this stirs some discussion and gets this moving:

> Should there be some additional message in the error about this being a panic turned error?
Or do we just produce the exact message the panic would produce?

Currently, the error message is like this ( https://play.rust-lang.org/?version=nightly&mode=debug&edition=2015&gist=1083e1ce78d0ca89e95fb5c3fc733cf5 ):
```
error[E0080]: evaluation of constant value failed
 --> src/main.rs:5:5
  |
5 |     panic!("panic while evaluating panic_in_const");
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the evaluated program panicked at 'panic while evaluating panic_in_const', src/main.rs:5:5
  |
  = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)

error: any use of this value will cause an error
 --> src/main.rs:8:15
  |
8 | const C: () = panic_in_const();
  | --------------^^^^^^^^^^^^^^^^-
  |               |
  |               referenced constant has errors
  |
  = note: `#[deny(const_err)]` on by default

error: aborting due to 2 previous errors
```

I think the error message could be better:
* `note: this error originates in a macro` seems like needless thing to say in this case.
* I didn't know about the knob that allows to alternate between "compilation fail right away" and "compilation fail upon a use of the constant". This seems like an useful thing (for macros, at least), and without the second error message, the users wouldn't learn about it. However, when testing it a bit, I can't make heads or tails how it actually works. (For example: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2015&gist=fdc4e6f2212a255b41497781e99c4cea )
  * Is it a module-level attribute?
  * Is it an attribute on the const fn?
  * Is it an attribute on the const item?
  * Seems buggy: trying allowing any of these, I can't make it compile, although I don't use the const value.
    * If that's not the point, then the error message is unclear and needs to be better.
    * What is the point, then?

> This change becomes really useful if Result::unwrap and Option::unwrap become const fn, doing both in one go might be a good idea.
* If and match are going to be stabilized, so these APIs are getting unblocked. I think we should go and stabilize their constness as a part of this push.

**Comment by RalfJung** [comment]
Note that so far only panics with a constant string are even implemented. That suffices for `Option::unwrap`, but `Result::unwrap` also formats the error value, and thus is far out of reach of CTFE.

**Comment by golddranks** [comment]
Maybe `Result::unwrap` should be left out of the scope then. I half-expected `expect` not to require formatting, but apparently it prints both the expect message and the error value, so no luck there. It should still be possible to unwrap `Results` using custom functions, so the capability to do so stays valuable even if `unwrap` can't be stabilized right away.

**Comment by golddranks** [comment]
About the first unsolved question: "Or do we just produce the exact message the panic would produce?" I wonder if there's some intention there that panics could be used for bona fide custom-made error messages, and that's why there would be motivation to produce "just the exact message"? As for that, I don't think we should mix and use messages that might have originally meant for run-time error messages, in a different, compile-time context *without an explanation about that context*. What I mean is that the compiler should clearly say for the reason for the error something like "a constant expression panicked" and then tell the panic error message.

I think that there is need and demand for proper custom-made compiler error messages/warnings in libraries, if the library is able to detect that its API is misused etc. However, the mechanism for that shouldn't be the same than for what is normally meant for run-time panics. It might be that a panicking mechanism is fit for even that, but in that case, the object associated with the panic should be some structured object with an API designed for showing and formatting great compile-time error messages. For purposes of this feature, it should be out of scope.

And thus: I think that the answer to the question is: yes, there should be an error message or an explanation that "frames" the error as a panic that happened in a constant. But preferably something more lightweight than what the current message looks like.

**Comment by golddranks** [comment]
As for the documentation: I noticed that mentions about `const fn`s are entirely missing from the Reference, which seems to be the relevant place of documentation for feature like this. Have the earlier stabilization PR's been neglecting updating it, or maybe there is so much churn around const fn feature that it makes sense to update it in one go?

**Comment by memoryruins** [comment]
> mentions about const fns are entirely missing from the Reference

`const fn` is documented in the reference https://doc.rust-lang.org/nightly/reference/items/functions.html#const-functions, but the section is slightly out of date https://github.com/rust-lang/reference/issues/800

**Comment by golddranks** [comment]
@memoryruins: Thanks, I missed it.

More about the `const_err`: In this example using `TEST` inside another constant counts as "use". https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=928d36c50cceec31d97999d5ef05447a

This explains the behaviour I called earlier "seemingly buggy". Using an "erroneous" constant even inside another constant is an error, even though I expected the error propagation logic to work transitively. However, as this example that doesn't use panicking at all shows, the problem isn't directly related to allowing panicking in constants: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=702befa4f5e87e6f997307c8419cbd62

Rather, it's about the propagation and presentation of "erroneous" constants. But with allowing panicking making erroneous constants becomes more normal and expected than before, so it may need to be considered as an UX problem.

**Comment by abonander** [comment]
> and we need to document this feature

The link to the stabilization guide on Forge is broken in the top-level issue. Is there a new reference for writing feature documentation?

**Comment by mark-i-m** [comment]
@abonander Currently, the stabilization guide is here: https://rustc-dev-guide.rust-lang.org/stabilization_guide.html

DISCLAIMER: there are active discussions about where we want this kind of content to live long-term...

**Comment by abonander** [comment]
> I think that the answer to the question is: yes, there should be an error message or an explanation that "frames" the error as a panic that happened in a constant. But preferably something more lightweight than what the current message looks like.

Anything wrong with just switching the message placement? E.g. from this:

```
error[E0080]: could not evaluate static initializer
 --> src/main.rs:3:21
  |
3 | static FOO: usize = panic!("Hello, world!");
  |                     ^^^^^^^^^^^^^^^^^^^^^^^ the evaluated program panicked at 'Hello, world!', src/main.rs:3:21
  |
  = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)
```

To this:

```
error[E0080]: Hello, world!
 --> src/main.rs:3:21
  |
3 | static FOO: usize = panic!("Hello, world!");
  |                     ^^^^^^^^^^^^^^^^^^^^^^^ panic in static initializer
  |
  = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)
```

I don't think there's any ambiguity here. The error is formatted as a compiler-error and the context message explains why it was a compiler error.

**Comment by oli-obk** [comment]
I like it. It's radical but powerful.

**Comment by RalfJung** [comment]
Yeah, `error[E0080]: could not evaluate static initializer` is not a terribly useful initial error. IMO we should change that for non-panic errors, too. In Miri, I changed errors to start like
```
error: unsupported operation: `clock_gettime` not available when isolation is enabled
```

**Comment by abonander** [comment]
> I like it. It's radical but powerful. It will just be fun to watch how ppl will use it to emit fake compiler errors by inserting newlines and color characters in their error messages, causing complete confusion (they could have done so with the way we emit messages currently, too, so this is nothing new, they can just do it more conveniently).

@oli-obk is that potentially a concern, if people begin to rely on exactly _where_ the compiler emits their const panic message? I guess we can easily disclaim that: "the exact rendering of the panic message is not considered stable" or something like that.

**Comment by Lokathor** [comment]
That seems like an unreasonable fear. People can use many parts of the language poorly, and we trust users to avoid crates that give them a bad experience.

But also, yes, the exact rendering should be explicitly classified as a "debug message" and thus subject to potential change without it being a "breaking change".

**Comment by mark-i-m** [comment]
Is there a reason not to guarantee that panics in constants will display this way? It would be kind of awesome to have a way for people to generate custom error messages...

**Comment by Lokathor** [comment]
I would like to not guarantee a specific display because I would like rustc to one day be able to detect color change ansi control codes and strip them out of the output when color output is turned off or automatically detected to be off.

And in general, specifying the exact nature of a debug message just isn't the usual for Rust.

I think saying "the panic message is shown to the user in some way as a compile error" is specific enough.

**Comment by oli-obk** [comment]
Even if we don't want to lock ourselves into anything, giving a reasonable default, like the one proposed by @abonander, seems like a good idea.

**Comment by rodrimati1992** [comment]
Could the format string allow specifying the names of the printed variables like in [RFC 2795](https://github.com/rust-lang/rfcs/blob/master/text/2795-format-args-implicit-identifiers.md) ? That way one can use only some of the variables in scope,and put them in any order.

**Comment by RalfJung** [comment]
Currently const-panics only allow string constants, no format string at all. That is still way off at this point I am afraid.

The format string syntax will be exactly the same as for run-time formatting.

**Comment by CodesInChaos** [comment]
Should the error include a stack trace? Otherwise trouble shooting panics that happened deep inside a const fn might be difficult to diagnose. (I don't think this is needed at release, but should be considered as a later enhancement)

**Comment by oli-obk** [comment]
We already emit a stack trace: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=b0f5cc58e32119df1f116595783db08e

```
8 |     [()][x]
  |     ^^^^^^^
  |     |
  |     index out of bounds: the len is 1 but the index is 4
  |     inside `foo` at src/main.rs:8:5
  |     inside `bar` at src/main.rs:4:5
  |     inside `FOO` at src/main.rs:1:17
```

**Comment by AaronKutch** [comment]
There is a workaround for this issue now in stable 1.46 that indexes outside of an array when we want to panic:
```
// reduced example for my usecase that involves wrappers around `NonZeroUsize`
pub const fn bw(width: usize) -> NonZeroUsize {
    ["Tried to construct an invalid BitWidth of 0 using the `apint::bw` function"][(width == 0) as usize];
    unsafe {NonZeroUsize::new_unchecked(width)}
}
```
When executed at compile time, it will show the message in the string, but unfortunately the runtime error is unhelpful.

**Comment by rodrimati1992** [comment]
@AaronKutch 
But this compiles in Rust 1.33
https://rust.godbolt.org/z/fThezn
```rust
use std::num::NonZeroUsize;

// reduced example for my usecase that involves wrappers around `NonZeroUsize`
const fn bw(width: usize) -> NonZeroUsize {
    ["Tried to construct an invalid BitWidth of 0 using the `apint::bw` function"][(width == 0) as usize];
    unsafe {NonZeroUsize::new_unchecked(width)}
}

pub static FOO: NonZeroUsize = bw(30);
```
Also, replacing the `30` with a `0` errors with index out of bounds as expected.

**Comment by AaronKutch** [comment]
I should have stated that it probably compiled before the current stable, but I did not know it would compile that long ago.

**Comment by rodrimati1992** [comment]
I noticed that you're required to pass a string literal to `core::panic`, and otherwise you need to use `std::panic` to panic with a `const S: &str = ....;`, is this going to be fixed soon?

I made a crate for doing compile-time formatting, and find it kind of annoying that the `assertc` macro for compile-time assertions won't be usable with `#![no_std]` crates once it's released.
 https://github.com/rodrimati1992/const_format_crates/tree/assertions#formatted-const-panics

**Comment by RalfJung** [comment]
@rodrimati1992 if you are asking whether it will be possible to use format strings for const panics, the answer is -- not for a long while. Format strings are extremely complicated.

The [const-eval skill tree](https://rust-lang.github.io/const-eval/) show the dependency chain to get panics with formatting to work.

**Comment by rodrimati1992** [comment]
I am asking about panicking in a `#[no_std]` crate with a `&'static str` `const`ant.
Example: the `S` in `const S: &str = const_format::formatc!("{:?}", Point{x: 3, y: 5});` (this is not pseudocode).

**Comment by RalfJung** [comment]
> probably because nobody anticipated that I would make my own formatting machinery that works before const allocations or traits in const fns do.

Yeah, probably.^^ I was not aware this is a thing, impressive.

But, what do you need to make this work `no_std`? Again, a self-contained example would help understand the limitation here.

**Comment by rodrimati1992** [comment]
Well, a self contained example would be:
```rust
#![feature(const_panic)]
//#![no_std]

use core::mem;

// Let's assume that FOO and BAR come from a dependency
const FOO: (usize, &str) = (mem::size_of::<usize>(), "foo is smaller than bar");
const BAR: (usize, &str) = (mem::size_of::<u32>(), "bar is smaller than foo");

const _: () = {
    if FOO.0 < BAR.0 {
        panic!(FOO.1)
    } else if FOO.0 > BAR.0  {
        panic!(BAR.1)
    }
};
```
Uncomment the `#![no_std]` line and you'll get errors about how it tries to do something non-const.

**Comment by lzutao** [comment]
This is because the different expansions of panic! in std and core.
In std, panic! is expanded to `::std::rt::begin_panic`, in core it is `::core::panicking::panic_fmt`.
You might want to use `core::panicking::panic` instead of panic!.

**Comment by RalfJung** [comment]
It's not `panic_fmt` though since no format string is used. The problem is `::core::panicking::panic`.

Probably this line needs an exception for that function:

https://github.com/rust-lang/rust/blob/c3364780d2cfddfe329f62a3ec138fd4f9a60e27/compiler/rustc_mir/src/transform/check_consts/mod.rs#L56

**Comment by lzutao** [comment]
> It's not `panic_fmt` though since no format string is used

It is what I get from `expanded`: https://rust.godbolt.org/z/rznTzh

**Comment by RalfJung** [comment]
Oh right, these are not string literals... I am actually surprised this works in `std::panic!`.

Also what is this:
https://github.com/rust-lang/rust/blob/c3364780d2cfddfe329f62a3ec138fd4f9a60e27/library/core/src/macros/mod.rs#L9
I never saw `literal` as a macro argument class before.^^

**Comment by RalfJung** [comment]
Okay so this works with both the core and std macro:
```rust
#![feature(const_panic)]

const FOO:() = std::panic!("hello");
const BAR:() = core::panic!("hello");
```
AFAIK that is the only part that is *intended* to work. The fact that `std::panic!` works with more arguments is, I think, an accident... Cc @oli-obk

**Comment by RalfJung** [comment]
Oh, this was actually broken by https://github.com/rust-lang/rust/pull/74056. Before that PR, `core::panic!` with one argument expanded to `::panicking::panic` and that worked curing const-time. It seems we didn't have a test like this though:
```rust
#![feature(const_panic)]

const MSG: &str = "hello";

const FOO:() = std::panic!(MSG);
const BAR:() = core::panic!(MSG);
```
So the regression went unnoticed.

**Comment by m-ou-se** [comment]
Since we're talking about panics on non-literals; this currently gives an ICE:
```rust
#![feature(const_panic)]

const _: () = panic!(123);
```

**Comment by RalfJung** [comment]
> Since we're talking about const panics on non-literals; this currently gives an ICE:

Yup, that's https://github.com/rust-lang/rust/issues/66693.

**Comment by tarcieri** [comment]
> Okay so this works with both the core and std macro:

```rust
#![feature(const_panic)]

const FOO:() = std::panic!("hello");
const BAR:() = core::panic!("hello");
```

Curious how hard it would be to stabilize this much. It's really all I need for now. [I've gone ahead and used the array hack to achieve the same effect](https://github.com/RustCrypto/utils/pull/76/files#diff-191d8d9b8b6b646e67653db61f347ef9R59), but it'd be nice to have something which provides a reasonable panic message when used in non-`const` contexts as well.

**Comment by RalfJung** [comment]
@rodrimati1992 
> I noticed that one is required to pass a string literal to core::panic, needing to use std::panic to panic with a const S: &str = ....;, is this going to be fixed soon?

So after having done the investigation here and at https://github.com/rust-lang/rust/issues/67984#issuecomment-687611718, this is very far from an easy issue to fix. I see no good way to resolve the problem. This is technically regression, but of a nightly-only feature, so probably not a strong enough argument to revert anything, given the arguments in favor of the change that caused the regression.

@tarcieri 
One blocking problem is adjusting the const checker to ensure that the argument is indeed a string literal, thus avoiding the ICE. I'm afraid that will break @rodrimati1992's approach; we could still allow that under a separate flag.

Once that's done, I think there was still some discussion around how exactly to print the panic message to the user, but I see no major blockers. I'd prefer to have at least somewhat settled promotion though before we stabilize more ways in which CTFE can fail (Cc https://github.com/rust-lang/const-eval/issues/53), but that is probably asking too much.

**Comment by m-ou-se** [comment]
The const non-literal core::panic can be fixed by turning that case also into a lang item. Working demo/patch: https://github.com/rust-lang/rust/compare/master...fusion-engineering-forks:core-const-panic-str

(Not sure if adding a lang item just for this is worth it though.)

**Comment by RalfJung** [comment]
Ah good point, we can separate that case from the more general `panic_fmt`.

**Comment by rodrimati1992** [comment]
The reason I am not using `core::panicking::panic` is that It's not even unstably const, so I don't know if it's only intended to be used at compile-time by the `core::panic` macro.

**Comment by RalfJung** [comment]
@rodrimati1992 that function is an unstable implementation detail that no external crates should use.

**Comment by josephlr** [comment]
I ran into the following weird situation:

The following code only needs `#![feature(const_panic)]`:
```rust
const A: () = assert!(1 == 1);
const B: () = panic!("Oh noes");
```

but this code needs `#![feature(const_fn)]` and `#![feature(const_panic)]`:
```rust
const fn foo() {
    assert!(1 == 1);
}
```

Is this intentional, or is it related to the https://github.com/rust-lang/rust/pull/74056 issue?

**Comment by RalfJung** [comment]
@josephlr Looks like `min_const_fn` does not accept panics even with `const_panic` enabled. Not sure if that is deliberate -- @oli-obk @ecstatic-morse do you know?

**Comment by josephlr** [comment]
I thought `core::panicing::panic` needed to be marked with `rustc_const_unstable`, but that didn't work.

**Comment by RalfJung** [comment]
The panic functions are non-const. They are subject to a [special exception in the const checks](https://github.com/rust-lang/rust/blob/c3364780d2cfddfe329f62a3ec138fd4f9a60e27/compiler/rustc_mir/src/transform/check_consts/mod.rs#L56), and then during const-evaluation the interpreter notices when these functions are called, and stops interpretation (so their bodies do not even matter).

**Comment by josephlr** [comment]
@RalfJung https://github.com/rust-lang/rust/pull/76602 fixes the issue, thanks for the tips!

**Comment by jonas-schievink** [comment]
> but this code needs `#![feature(const_fn)]` and `#![feature(const_panic)]`:
> 
> ```rust
> const fn foo() {
>     assert!(1 == 1);
> }
> const fn bar() {
>     panic!("hi");
> }
> ```

That looks expected, since `foo` uses `PartialEq`, and trait methods aren't stable in CTFE (https://github.com/rust-lang/rust/issues/67792).

Just `bar` alone compiles fine just with `#![feature(const_panic)]`.

**Comment by oli-obk** [comment]
Both of these are fine with just `const_panic`, because `==` on primitive integers does not go through `PartialEq`. You may have been thinking of `assert_eq`, which has a whole host of problems like `Debug` impls and such

**Comment by jonas-schievink** [comment]
Ah, right

**Comment by SergioBenitez** [comment]
I would love to see this move toward stabilization. If the issue description is up-to-date, we should be ready to do so:

> Should there be some additional message in the error about this being a panic turned error? Or do we just produce the exact message the panic would produce?

Given that there is no stability guarantee regarding error messages, this should not preclude stabilization. Nevertheless, I would suggest that a `panic!()` evaluated in a `const` context act just like a `compile_error!()` in a `macro` context, for consistency.

> This change becomes really useful if `Result::unwrap` and `Option::unwrap` become `const fn`, doing both in one go might be a good idea.

This is a nice idea in theory but is orthogonal to stabilizing this feature. 

> **Blockers:**
> * ICE: #66693

This, the only real blocker, is now resolved.

**Comment by RalfJung** [comment]
> Given that there is no stability guarantee regarding error messages, this should not preclude stabilization. Nevertheless, I would suggest that a panic!() evaluated in a const context act just like a compile_error!() in a macro context, for consistency.

There was a lot of discussion around this at https://github.com/rust-lang/rust/pull/79695.

However, I agree that this does not have to be a blocker.

**Comment by jhpratt** [comment]
Any opposition to stabilizing this as-is? Sounds like all blockers have been resolved, and as mentioned the exact formatting of the error message is not guaranteed to remain unchanged.

**Comment by RalfJung** [comment]
> Any opposition to stabilizing this as-is?

Not from my side.

**Comment by nikomatsakis** [comment]
Nominating for an upcoming @rust-lang/lang meeting to discuss potential stabilization. I would encourage however that if someone wants to see this stabilized, the next step is to write a [stabilization report](https://rustc-dev-guide.rust-lang.org/stabilization_guide.html#write-a-stabilization-report).

**Comment by Mark-Simulacrum** [comment]
Discussed briefly in lang team meeting - generally positive, looking forward to reviewing a stabilization report.

**Comment by JohnTitor** [comment]
# Stabilization report

## Summary

This feature allows panicking within constants and reporting a custom message as a compile-time error.

The following code will become valid on stable:

```rust
const fn foo(val: u8) {
    assert!(val == 1, "val is not 1");
}

const MSG: &str = "custom message";

const _: () = std::panic!(MSG);
const _: () = panic!("custom message");
```

The message looks like:

```
error: any use of this value will cause an error
 --> src/main.rs:4:15
  |
4 | const _: () = std::panic!(MSG);
  | --------------^^^^^^^^^^^^^^^^-
  |               |
  |               the evaluated program panicked at 'custom message', src/main.rs:4:15
  |
  = note: `#[deny(const_err)]` on by default
  = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
  = note: for more information, see issue #71800 <https://github.com/rust-lang/rust/issues/71800>
  = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)
```

Motivation for this RFC can be found here: https://rust-lang.github.io/rfcs/2345-const-panic.html#motivation

More examples can be found here: https://github.com/rust-lang/rust/blob/673d0db5e393e9c64897005b470bfeb6d5aec61b/src/test/ui/consts/const-eval/const_panic.rs

## Edge cases & tests

- https://github.com/rust-lang/rust/blob/673d0db5e393e9c64897005b470bfeb6d5aec61b/src/test/ui/consts/const-eval/panic-never-type.rs
- https://github.com/rust-lang/rust/blob/673d0db5e393e9c64897005b470bfeb6d5aec61b/src/test/ui/consts/const-eval/panic-assoc-never-type.rs
- https://github.com/rust-lang/rust/blob/673d0db5e393e9c64897005b470bfeb6d5aec61b/src/test/ui/consts/issue-66693.rs
- https://github.com/rust-lang/rust/blob/673d0db5e393e9c64897005b470bfeb6d5aec61b/src/test/ui/consts/control-flow/short-circuit-let.rs

## Some questions which should be resolved in the future

We have some "unresolved questions" but they aren't an actual blocker.


- Should there be some additional message in the error about this being a panic turned error?
   Or do we just produce the exact message the panic would produce?
- This change becomes really useful if `Result::unwrap` and `Option::unwrap` become `const fn`, doing both in one go might be a good idea.

See the brief summary comment here: https://github.com/rust-lang/rust/issues/51999#issuecomment-790543959

**Comment by jonas-schievink** [comment]
> This change becomes really useful if `Result::unwrap` and `Option::unwrap` become `const fn`, doing both in one go might be a good idea.

`Result::unwrap` requires const trait impls, but I think `Option::unwrap` requires nothing besides panicking

**Comment by programmerjake** [comment]
While we're at it, we should go through `Result` and `Option` and `const`ify all their methods that don't depend on unstable features (such as `unwrap_or`).

**Comment by camelid** [comment]
I'm wondering: should we make a panic in a constant a hard error (instead of using the `const_err` lint)? Since it hasn't been stabilized yet, I think it's okay to make it a hard error. Or would that be confusing because it would introduce inconsistency?

**Comment by jhpratt** [comment]
@programmerjake I went through all non-`const` `Result` methods a bit ago. Stabilizing const panic does not permit _a single method_ to be `const fn` on stable by itself — most require const trait impls. See #82814 for reference.

`Option` isn't quite as bad, as there's not an "other" value that would need dropping.

**Comment by RalfJung** [comment]
> I'm wondering: should we make a panic in a constant a hard error (instead of using the const_err lint)? Since it hasn't been stabilized yet, I think it's okay to make it a hard error. Or would that be confusing because it would introduce inconsistency between different const errors?

Hm... I am on the fence about this. On the one hand our CTFE error reporting code is already a mess and this would probably make it worse; on the other hand it is a small step towards making `const_err` a hard error which I like.

**Comment by nikomatsakis** [comment]
Can someone summarize the differences in behavior under discussion here? I am not following closely enough to really tell.

**Comment by CodesInChaos** [comment]
I find it a bit weird to allow non-literals as argument to const panic, when that functionality will be disabled in Rust 2021.

I'd rather restrict it to literals, and look into allowing trivial patterns as next step (all format arguments are `&str` and the pattern doesn't contain any format options), which would aloow functions like `Option::Expect` to be const without relying on traits.

**Comment by camelid** [comment]
@nikomatsakis 

> Can someone summarize the differences in behavior under discussion here? I am not following closely enough to really tell.

If you are referring to the discussion about const panic in relation to `const_err`, then the difference would be whether const panic emits a deny-by-default `const_err` *lint*, or its own hard error. I believe the plan is to eventually turn `const_err` into a hard error, and because const panic is currently an unstable feature, we could make it a hard error now, rather than waiting until the edition after 2021.

**Comment by Lokathor** [comment]
I *think* the main reason to allow a const error currently is because of the lack of const panic, so if const panic was stable then we're probably good.

or are there other reasons to keep allowing const errors that can't just be a const panic instead?

**Comment by camelid** [comment]
> I _think_ the main reason to allow a const error currently is because of the lack of const panic, so if const panic was stable then we're probably good.
> 
> or are there other reasons to keep allowing const errors that can't just be a const panic instead?

Could you link to an example of how `allow(const_err)` would be used due to the lack of const panic? I've only used simple constants so I don't have knowledge of what kind of issues come up.

**Comment by Lokathor** [comment]
Oh, turns out I'm thinking `allow(unconditional_panic)`, my mistake.

**Comment by SergioBenitez** [comment]
> @nikomatsakis 
> 
> 
> 
> > Can someone summarize the differences in behavior under discussion here? I am not following closely enough to really tell.
> 
> 
> 
> If you are referring to the discussion about const panic in relation to `const_err`, then the difference would be whether const panic emits a deny-by-default `const_err` *lint*, or its own hard error. I believe the plan is to eventually turn `const_err` into a hard error, and because const panic is currently an unstable feature, we could make it a hard error now, rather than waiting until the edition after 2021.

For the record, I don't think anything _but_ a hard error makes sense. Otherwise, the panic _will_ unequivocally occur at runtime, even though the compiler would have trivially prevented it, so why shouldn't it?

**Comment by camelid** [comment]
> For the record, I don't think anything _but_ a hard error makes sense. Otherwise, the panic _will_ unequivocally occur at runtime, even though the compiler would have trivially prevented it, so why shouldn't it?

Unless I'm misunderstanding, using `allow(const_err)` will still cause compilation to fail, just with a confusing error message:

```rust
#![feature(const_panic)]
#![allow(const_err)]

const X: usize = panic!();

fn main() {
    println!("{}", X);
}
```

```
error[E0080]: evaluation of constant value failed
 --> src/main.rs:7:20
  |
7 |     println!("{}", X);
  |                    ^ referenced constant has errors

error: aborting due to previous error
```

**Comment by CodesInChaos** [comment]
With `allow(const_err)` you get a compiler error if the constant is used, but it's compile if it's unused.

Without `allow(const_err)` you get a compiler error even if the constant is unused.

**Comment by Mark-Simulacrum** [comment]
> For the record, I don't think anything but a hard error makes sense. Otherwise, the panic will unequivocally occur at runtime, even though the compiler would have trivially prevented it, so why shouldn't it?

Hm. It seems plausible that the program could do a number of useful things before it reaches the panic, and just like in regular code adding an unimplemented!() and continuing to work on other pieces is useful, the same seems true of const-expr. I think it makes sense for the compiler to warn (and maybe even error) by default on such cases, but entirely refusing to compile code that will eventually fail at runtime doesn't seem quite right.

**Comment by DoumanAsh** [comment]
Panic should evaluate to compile error only at compile time, if `const` expression is not evaluated at compile time compiler wouldn't be able to detect it at all so I don't think it would be possible to detect such panic and error at all?

**Comment by RalfJung** [comment]
> For the record, I don't think anything but a hard error makes sense. Otherwise, the panic will unequivocally occur at runtime, 

No, const-panics will never occur at runtime. As @CodesInChaos said, the only difference is for *unused* constants. The following will always fail to compile:
```rust
const FOO: i32 = panic!();

fn main() {
  let _x = FOO;
}
```

>  but entirely refusing to compile code that will eventually fail at runtime doesn't seem quite right.

That's not what happens; it will fail to build since a constant has failed to evaluate. We *guarantee* that code that syntactically mentions a `const` will fail to build if that constant fails to evaluate, and there is unsafe code relying on that for safety.

**Comment by AaronKutch** [comment]
There should certainly be a hard error by default whenever any constant panics. I have code that relies on that being the case even in unused constants (which make sure that certain invariants are being kept and stops compilation in case of some problem). There could be an attribute that changes the default behavior for unused constants, but never some global flag or other externally controllable change that would allow my crate to compile if its unused constants panicked.

**Comment by CodesInChaos** [comment]
Making rules less restrictive is no breaking change, so I suggest emitting a hard error initially and possibly loosening it up later.

**Comment by AaronKutch** [comment]
I want to make sure that the default in my crate _never ever_ changes unless I explicitly add an attribute in it.

**Comment by RalfJung** [comment]
>  I have code that relies on that being the case even in unused constants

What do you mean by "unused constant"? Note that for some unused constants, this is impossible, e.g.
```rust
trait Foo {
  const C: i32;
}

impl<T> Foo for Vec<T> {
  const C: i32 = panic!();
}
```
Here, `C` can only be evaluated once a particular `T` is picked. So such a panic-in-unused-const like the above will not and cannot cause a compilation error -- this is generally true for associated consts.

For global `const`, on the other hand, we *do* eagerly evaluate all of them, but unless they *syntactically* appear in actual code, there is no guarantee that anything will cause an error, and you would be incorrect to rely on this. See https://github.com/rust-lang/rust/issues/71800.

**Comment by ojeda** [comment]
> We guarantee that code that syntactically mentions a const will fail to build if that constant fails to evaluate (even if the value of the constant is irrelevant for the code or only occurs in dead code)

Currently, it compiles for dead code, e.g.:

```rust
#[allow(const_err)]
const C: u8 = unimplemented!();

fn f() -> u8 {
    C
}

pub fn main() {
}
```

Nevertheless, I agree a hard error seems best.

> There should certainly be a hard error by default whenever any constant panics. I have code that relies on that being the case even in unused constants (which make sure that certain invariants are being kept and stops compilation in case of some problem).

Using them as `static_assert`? :)

> There could be an attribute that changes the default behavior for unused constants, but never some global flag or other externally controllable change that would allow my crate to compile if its unused constants panicked.

So something like the attribute in the example above, but without being a lint, right?

**Comment by RalfJung** [comment]
> Currently, it compiles for dead code, e.g.:

Hm, yes, I guess we don't actually generate code for `f` so we'd never actually evaluate that `const`. The guarantee I talked about is for functions that actually get called.

**Comment by AaronKutch** [comment]
It makes sense to me that
```
trait Foo {
  const C: i32;
}

impl<T> Foo for Vec<T> {
  const C: i32 = panic!();
}
```
Couldn't cause a hard error until a T is picked, because the panic could depend on properties of `T`. However, I strongly feel that once a concrete constant exists, it should always have static assertion like behavior by default. In case of examples like making the constant return `unimplemented!()`, `#[allow(const_err)]` could be added. It is taking the conservative approach but allowing an escape hatch.

**Comment by RalfJung** [comment]
> However, I strongly feel that once a concrete constant exists, it should always have static assertion like behavior by default. In case of examples like making the constant return unimplemented!(), #[allow(const_err)] could be added. It is taking the conservative approach but allowing an escape hatch.

That sounds like the current behavior then -- but note that this has not been RFC'd so it is not clear how much of this is a guarantee.

**Comment by AaronKutch** [comment]
This seems like an edge case that RFC 2345 didn't specify. Can RFCs be amended, or is there some place to add small clarifications?

**Comment by Lokathor** [comment]
RFCs are usually like "log of what we planned to do at the time" rather than "spec of how it works". Usually an old RFC will not be amended directly, a new RFC would just be introduced and passed.

**Comment by nikomatsakis** [comment]
New issue: https://github.com/rust-lang/rust/issues/85194

**Comment by jhpratt** [comment]
Now that #86998 is merged, this should be good to stabilize, correct?

**Comment by oli-obk** [comment]
I'm still unhappy that in 2018 we can do `panic!(some_computed_str)` but in 2021 we can't do the equivalent, but it seems better to only allow literal strings for now than to not allow any custom panicking. Imo this is purely a T-libs issue, but I'll let @rust-lang/wg-const-eval weigh in on that before escalating to T-libs for stabilization.

**Comment by RalfJung** [comment]
> it seems better to only allow literal strings for now than to not allow any custom panicking

:+1: as long as this is enforced properly(i.e. everything accepted in 2018 on stable also works in 2021).

**Comment by oli-obk** [comment]
it is not yet. Not sure how doable that is, it may be easier to just figure out how to do `panic!("{}", some_computed_str)` on all editions.

**Comment by abonander** [comment]
For our planned usage of this feature in SQLx we would really prefer either support for formatting or being able to compute panic strings. We can _sorta_ work around it if the panic strings need to be literals but it won't be the greatest.

**Comment by clarfonthey** [comment]
One thing that I noticed which kind of sucks: this being unstable means that you can't use `todo!()` in consts. Although I understand not wanting to stabilise that so that people don't use it instead of panic.

**Comment by oli-obk** [comment]
Yes this is intentional. Getting that to work was actually a blocker for stabilization.

**Comment by tmandry** [comment]
Is there an issue tracking support for non-`&str` panic arguments in consts?

Right now the compiler just gives an error (implemented in #80734) "argument to `panic!()` in a const context must have type `&str`", but there's no error code or other explanation text.

The RFC doesn't mention this limitation.

**Comment by oli-obk** [comment]
There's no special tracking for it. It's essentially just blocked on the fact that you can't implement Display or Debug in a way that is callable in a const context.

**Comment by rodrigocfd** [comment]
Damn, I just stumbled upon a situation where a I need a fallible enum construction, [like this one](https://rust-lang.github.io/rfcs/2345-const-panic.html#guide-level-explanation):

```rust
const fn parse_bool(s: &str) -> bool {
    match s {
        "true" => true,
        "false" => false,
        other => panic!("`{}` is not a valid bool", other), // CAN'T DO THAT!
    }
}
```

Time to refactor a huge workaround 😞

I really hope this issue can be fixed, I have so many other situations where it would make my design correct.

**Comment by golddranks** [comment]
@rodrigocfd You can already panic in const fns, you just can't use string formatting machinery yet. Remove the `{}` and `, other` and the panic part compiles. However, `&str` matching doesn't seem to work in const fns yet, as that requires const-agnostic traits.

Also a side note, panicing is a bit suspect strategy if you are _expecting_ the matching to fail with some inputs.

This is a closed tracking issue for panics in const and not meant for general discussion, if you need help within the current capabilities of the compiler, try https://users.rust-lang.org or https://www.reddit.com/r/rust/

**Comment by rodrigocfd** [comment]
@golddranks thank you.
